janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphTraversalUtil.java

```diff
@@ -103,45 +109,140 @@ public static JanusGraphTransaction getTx(Traversal.Admin<?, ?> traversal) {
+            } else {
+                return Optional.of(currentStep); // place JanusGraphMultiQueryStep at the stat of the local traversal
+            }
+        } else if (previousStep instanceof NoOpBarrierStep) {
+            return Optional.of(previousStep);
+        } else {
+            return Optional.of(currentStep);
+        }
+    }
+
+    /**
+     * Checks whether this step is a traversal parent for which a preceding <code>JanusGraphMultiQueryStep</code>
+     * can have a positive impact on the step's child traversals.
+     * @param step The step to be checked.
+     * @return <code>true</code> if the step's child traversals can possibly benefit of a preceding
+     *         <code>JanusGraphMultiQueryStep</code>, otherwise <code>false</code>.
+     */
+    public static boolean isMultiQueryCompatibleParent(Step<?, ?> step) {
+        return MULTIQUERY_COMPATIBLE_PARENTS.stream().anyMatch(stepClass -> stepClass.isInstance(step));
     }
-    private static void getMultiQueryCompatibleStepsFromChildTraversal(Traversal.Admin<?,?> childTraversal, Step parentStep, Set<Step> multiQueryCompatibleSteps) {
-        Step firstStep = childTraversal.getStartStep();
-        while (firstStep instanceof StartStep || firstStep instanceof SideEffectStep) {
-            // Want the next step if this is a side effect
-            firstStep = firstStep.getNextStep();
+
+    /**
+     * Checks whether a step can profit of a preceding <code>JanusGraphMultiQueryStep</code>.
+     * @param step The step for which the condition is checked.
+     * @return <code>true</code> if the step is either a <code>MultiQueriable</code> or is a MultiQuery compatible
+     *         parent step.
+     * @see MultiQueriable
+     */
+    public static boolean isMultiQueryCompatibleStep(Step<?, ?> step) {
+        return step instanceof MultiQueriable || isMultiQueryCompatibleParent(step);
+    }
+
+    /**
+     * This method closely matches the behavior implemented in <code>LazyBarrierStrategy</code> which ensures that
+     * no <code>NoOpBarrierStep</code>s are inserted if path labels are required. Since the same limitation applies
+     * to <code>JanusGraphMultiQueryStep</code>s, the definition of legal positions for this step is the same.
+     * @param step The step which follows the JanusGraphMultiQueryStep to be inserted.
+     * @return <code>true</code> if no path labels are required for this position, otherwise <code>false</code>.
+     * @see org.apache.tinkerpop.gremlin.process.traversal.strategy.optimization.LazyBarrierStrategy
+     */
+    public static boolean isLegalMultiQueryPosition(Step<?, ?> step) {
+        if (step.getTraversal().getTraverserRequirements().contains(TraverserRequirement.PATH)) return false;
+        boolean labeledPath = false;
+        Step currentStep = step.getTraversal().getStartStep();
+        while (currentStep.hashCode() != step.hashCode()) { // Codacy fails without hashCode()
```

Comment:
Does anybody have a better solution here? What I actually want is `currentStep != step`, but Codacy doesn't allow that and suggests to use `equals()`. However, multiple steps within a traversal can be equal without being _the same_.

Reply:
If two objects are equal according to equals() method, then their hashCode() must be the same too.

Reply:
I see now that the implementation of `hashCode()` actually only checks the step's class and labels. So I will revert that to simply use `currentStep != step` and let Codacy fail. The only other idea that comes to my mind is to resolve the index `step` in the traversal and compare them by their position.
I don't understand why this code results in a warning in the first place. I mean there's a reason both `equals()` and `==` exist.

Reply:
I am just trying to judge codacy's choice: if you use "==" here without explicitly explaining why, then very likely developers in the future will treat this as a bug.

Reply:
After having a closer look at the implementation of `equals()` for traversal steps, I found out that the issue that I had where `==` and `equals()` behave differently can probably be solved by calling `equals(other, true)` instead. I wasn't aware of this method before.

Reply:
Sometimes Codacy wrongly fails. That shouldn't block anyone. Initially we wanted to set GH action from Codacy into "Warning" state but there is no such state in GitHub actions. The only way to know that there are suggestions from Codacy is to fail the action. Again, failed Codacy action means that there are some optional suggestions from Codacy. It doesn't mean they are required to be resolved.